Makefile

```diff
@@ -0,0 +1,47 @@
+PROJECT := $(dir $(abspath $(lastword $(MAKEFILE_LIST))))
+
+SRC := $(PROJECT)src
+TESTS := $(PROJECT)tests
+ALL := $(SRC) $(TESTS)
+
+export PYTHONPATH = $(PROJECT):$(PROJECT)/lib:$(SRC)
+
+update-dependencies:
+	uv lock -U --no-cache
+	uv pip compile -q --no-cache pyproject.toml -o requirements.txt
+	uv pip compile -q --no-cache --all-extras pyproject.toml -o requirements-dev.txt
+
+generate-requirements:
+	uv pip compile -q --no-cache $(PROJECT)pyproject.toml \
+		-o $(PROJECT)requirements.txt
+	uv pip compile -q --no-cache --all-extras $(PROJECT)pyproject.toml \
+		-o $(PROJECT)requirements-dev.txt
+
+lint:
+	uv tool run ruff check $(ALL)
+	uv tool run ruff format --check --diff $(ALL)
+
+fmt:
+	uv tool run ruff check --fix $(ALL)
+	uv tool run ruff format $(ALL)
+
+unit:
+	uv tool run --with-requirements $(PROJECT)requirements-dev.txt \
+		coverage run \
+		--source=$(SRC) \
+		-m pytest \
+		--ignore=$(TESTS)/integration \
+		--tb native \
+		-v \
+		-s \
+		$(ARGS)
+	uv tool run --with-requirements $(PROJECT)requirements-dev.txt coverage report
+
+integration:
+	uv tool run --with-requirements $(PROJECT)requirements-dev.txt \
```

Comment:
Can you explain how the uv.lock file interacts with these targets?  iiuc from the [docs](https://docs.astral.sh/uv/reference/cli), `uv tool run` doesn't consume the uv.lock file, so I think the real source of truth here is the `requirements*.txt` files and the lock file is informational.  afaict the only time `uv.lock` comes into play is if we're using environments (ex: `uv sync` will update an env but keep any constraints spelled out in the lock file)

Reply:
Yeh, in fairness we could probably use `uv run` here, which would in fact consume the lock file. In this case, the lock file is used to generate the requirements files, so either way we should get a deterministic set of dependencies.

Reply:
Since charmcraft doesn't consume the lock file, I'm on the fence on whether to include it here at all.  All we're really locking to is the requirements.txt.  

It would have value if there was a `pack` target here like [data platform's](https://github.com/canonical/mysql-k8s-operator/blob/a0d3948d6603a3b2aeb7015cf6af819c594aa936/tox.ini#L32) which generates the requirements.txt on the fly.  Otherwise, we have two sources of truth

Reply:
I don't think that's necessarily true. In fairness - I could actually just omit the requirements files based on the latest outcome from the WG discussing this, and just commit the lock file, and generate the requirements in the charmcraft.yaml. 

BRB... ;-)

Reply:
Fixed, I think. See [diff](https://github.com/canonical/parca-k8s-operator/compare/c42e1edcc253f4c5158a294d1b5fa01057884c66..4e70437c972e25b3b7046f7a5b24505e5c1975b7)

Reply:
Ah yes ok this I like a lot more.  iiuc anyone doing `git clone; charmcraft pack` will now get exactly the packages that were used in the recent uv.lock, which is perfect.  Now I really like using the uv.lock

I think there's still missing flags on the `uv run` commands though.  (disclaimer: **not** a strong uv user) `uv run`, when run in a project (aka something that has a pyproject.toml?) [modifies the environment](https://docs.astral.sh/uv/reference/cli/#uv-run) and partially updates the uv.lock (adds anything missing, but doesn't change anything that exists, I think?).  So atm if you add `numpy` to your pyproject.toml deps right now, doing `make test` will modify `uv.lock`, install numpy, and then run the tests. Probably not what we want.

I think we need `uv run --locked ...`, which will error if `uv.lock` needs to be updated (because something is missing, or a real dependency conflict).  That way `uv.lock` is only edited by `make generate-requirements`, and `make unit/integration` loudly break if there's an inconsistency

..I'm confused why there's also `uv run --frozen` though... I think `--frozen` will silently suppress some conflicts but not others?

I don't understand which flags you're missing though.

Reply:
> iiuc anyone doing git clone; charmcraft pack will now get exactly the packages that were used in the recent uv.lock

Exactly!

The other behaviour is intentional -- in order for something new to be added it would need to be added in the `pyproject.toml`, which is what we want I think...

In my opinion, if you add something to `pyproject.toml` as a dependency of your project, you'd want it to be added to the lock file and made available in the test environments -- otherwise you'd have to go through a bunch of extra steps to make stuff work.

It's not like you'd miss it in a PR either -- the lock file would change, and the pyproject.toml would be changed...

Reply:
ah so then we're down to the area of preferences :)

Yeah I see what you're saying.  My first instinct was that a lock file shouldn't change as a side effect, only as an explicit act.  But you're right that changing the pyproject.toml expresses intent.  Not that its a ton of extra work to manually update the lock (`make update-dependencies`), but it does require *some* additional work.

Playing with the scenarios: 
1. if uv.lock and pyproject.toml are *consistent* (no constraint in pyproject.toml is violated by uv.lock), then `uv run` uses exactly the versions in `uv.lock` (even if there's newer options available)
2. if uv.lock has a package that violates a constraint in pyproject.toml, `uv run` will update the `uv.lock`
3. if you add a dependency to pyproject.toml, `uv run` will install that package and update the `uv.lock`
4. if you remove a package from `pyproject.toml`, `uv run` **will not** remove it from `uv.lock`

I like the current configuration everything but (4) - (4) feels like it should break loudly. This is a minor difference though.  I lean toward "block me from mistakes" whereas yours is more "smooth development flow".  Both lgtm